File: Game/web/player_stats.js
export class PlayerStats {
  constructor() {
    this.reset();
  }

  reset(lives = 3) {
    this.score = 0;
    this.correctWords = 0;
    this.totalAttempts = 0;
    this.totalTypedChars = 0;
    this.totalCorrectChars = 0;
    this.combo = 0;
    this.bestCombo = 0;
    this.lives = lives;
    this.startedAt = 0;
    this.finishedAt = 0;
    this.lastMessage = "";
  }

  startRun() {
    this.startedAt = performance.now();
    this.finishedAt = 0;
  }

  finishRun() {
    this.finishedAt = performance.now();
  }

  registerAttempt(typed, target, success, timeLeft) {
    this.totalAttempts += 1;
    this.totalTypedChars += typed.length;
    this.totalCorrectChars += [...typed].reduce((count, char, index) => count + (char === target[index] ? 1 : 0), 0);

    if (success) {
      this.correctWords += 1;
      this.combo += 1;
      this.bestCombo = Math.max(this.bestCombo, this.combo);
      const scoreDelta = 100 + Math.floor(timeLeft * 45) + this.combo * 15;
      this.score += scoreDelta;
      this.lastMessage = this._successMessage();
      return scoreDelta;
    }

    this.combo = 0;
    this.lastMessage = ["Too Slow!", "Wrong!", "Focus up!"][(Math.random() * 3) | 0];
    return 0;
  }

  _successMessage() {
    if (this.combo >= 10) return `Combo x${this.combo}!`;
    if (this.combo >= 5) return `Combo x${this.combo}!`;
    if (this.combo >= 3) return "Perfect!";
    return ["Great!", "Nice!", "Clean!", "Sharp!"][(Math.random() * 4) | 0];
  }

  get accuracy() {
    if (this.totalTypedChars === 0) return 100;
    return Math.round((this.totalCorrectChars / this.totalTypedChars) * 1000) / 10;
  }

  get wpm() {
    if (!this.startedAt) return 0;
    const end = this.finishedAt || performance.now();
    const minutes = Math.max((end - this.startedAt) / 60000, 1 / 60);
    return Math.round((this.correctWords / minutes) * 10) / 10;
  }
}

File: Game/player_stats.py
from __future__ import annotations

from dataclasses import dataclass, field
from time import perf_counter
import random


@dataclass
class PlayerStats:
    score: int = 0
    correct_words: int = 0
    total_attempts: int = 0
    total_typed_chars: int = 0
    total_correct_chars: int = 0
    combo: int = 0
    best_combo: int = 0
    lives: int = 3
    started_at: float = 0.0
    finished_at: float = 0.0
    last_message: str = ""
    last_message_kind: str = "neutral"
    history: list[dict] = field(default_factory=list)

    def reset(self, lives: int) -> None:
        self.score = 0
        self.correct_words = 0
        self.total_attempts = 0
        self.total_typed_chars = 0
        self.total_correct_chars = 0
        self.combo = 0
        self.best_combo = 0
        self.lives = lives
        self.started_at = 0.0
        self.finished_at = 0.0
        self.last_message = ""
        self.last_message_kind = "neutral"
        self.history.clear()

    def start_run(self) -> None:
        self.started_at = perf_counter()
        self.finished_at = 0.0

    def finish_run(self) -> None:
        self.finished_at = perf_counter()

    def register_attempt(self, typed: str, target: str, success: bool, time_left: float) -> int:
        self.total_attempts += 1
        self.total_typed_chars += len(typed)
        self.total_correct_chars += sum(1 for typed_char, target_char in zip(typed, target) if typed_char == target_char)

        if success:
            self.correct_words += 1
            self.combo += 1
            self.best_combo = max(self.best_combo, self.combo)
            score_delta = 100 + int(time_left * 45) + (self.combo * 15)
            self.score += score_delta
            self.last_message = self._success_message()
            self.last_message_kind = "success"
        else:
            self.combo = 0
            score_delta = 0
            self.last_message = random.choice(["Too Slow!", "Wrong!", "Focus up!"])
            self.last_message_kind = "failure"

        self.history.append(
            {
                "typed": typed,
                "target": target,
                "success": success,
                "time_left": round(time_left, 2),
            }
        )
        return score_delta

    def _success_message(self) -> str:
        if self.combo >= 10:
            return f"Combo x{self.combo}!"
        if self.combo >= 5:
            return f"Combo x{self.combo}!"
        if self.combo >= 3:
            return "Perfect!"
        return random.choice(["Great!", "Nice!", "Clean!", "Sharp!"])

    @property
    def accuracy(self) -> float:
        if self.total_typed_chars == 0:
            return 100.0
        value = (self.total_correct_chars / self.total_typed_chars) * 100.0
        return round(value, 1)

    @property
    def wpm(self) -> float:
        if self.started_at <= 0.0:
            return 0.0
        end_time = self.finished_at if self.finished_at > 0.0 else perf_counter()
        elapsed_minutes = max((end_time - self.started_at) / 60.0, 1 / 60.0)
        words_per_minute = self.correct_words / elapsed_minutes
        return round(words_per_minute, 1)
